File: solutions/typescript/111-flood-fill.ts
/**
 * Problem 111: Flood Fill (LeetCode 733)
 * Difficulty: Easy
 * Language: TypeScript
 */
function floodFill(image: number[][], sr: number, sc: number,
                    color: number): number[][] {
  const orig = image[sr][sc];
  if (orig === color) return image;
  const fill = (img: number[][], r: number, c: number): number[][] => {
    if (r < 0 || r >= img.length || c < 0 || c >= img[0].length) return img;
    if (img[r][c] !== orig) return img;
    const updated = img.map((row, i) =>
      i === r ? row.map((v, j) => j === c ? color : v) : row
    );
    return [[-1,0],[1,0],[0,-1],[0,1]].reduce(
      (acc, [nr, nc]) => fill(acc, r+nr, c+nc), updated);
  };
  return fill(image, sr, sc);
}

File: solutions/elixir/111-flood-fill.ex
# Problem 111: Flood Fill (LeetCode 733)
# Difficulty: Easy
# Language: Elixir
# 
defmodule FloodFill do
  def fill(image, sr, sc, color) do
    orig = image |> Enum.at(sr) |> Enum.at(sc)
    if orig == color, do: image, else: do_fill(image, sr, sc, color, orig)
  end
  defp do_fill(img, r, c, color, orig) do
    rows = length(img); cols = length(hd(img))
    if r < 0 or r >= rows or c < 0 or c >= cols do img
    else
      if Enum.at(Enum.at(img, r), c) != orig do img
      else
        updated = List.update_at(img, r, &List.replace_at(&1, c, color))
        [{-1,0},{1,0},{0,-1},{0,1}]
        |> Enum.reduce(updated, fn {dr, dc}, acc ->
          do_fill(acc, r+dr, c+dc, color, orig) end)
      end
    end
  end
end
